controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		return ctrl.Result{}, err
 	}
 
-	if err := r.ReconcileNormal(rsCtx); err != nil {
+	result, err := r.ReconcileNormal(rsCtx)
+	if err != nil {
 		rsCtx.Logger.Error(err, "Failed to reconcile VirtualMachineReplicaSet")
 		return ctrl.Result{}, err
 	}
 
-	return ctrl.Result{}, nil
+	return result, nil
+}
+
+func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineReplicaSetContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineReplicaSet Deletion")
+
+	if controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
+		defer func() {
+			r.Recorder.EmitEvent(ctx.ReplicaSet, "Delete", nil, false)
+		}()
+
+		controllerutil.RemoveFinalizer(ctx.ReplicaSet, finalizerName)
+	}
+
+	return nil
+}
+
+// shouldExcludeVirtualMachine returns true if the VirtualMachine should be
+// filtered out, false otherwise.
+func shouldExcludeVirtualMachine(rs *vmopv1.VirtualMachineReplicaSet, vm *vmopv1.VirtualMachine) bool {
+	if metav1.GetControllerOf(vm) != nil && !metav1.IsControlledBy(vm, rs) {
+		return true
+	}
+
+	return false
 }
 
-func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineReplicaSetContext) error {
+// adoptOrphan sets the VirtualMachineReplicaSet as a controller OwnerReference
+// to the VirtualMachine.
+func (r *Reconciler) adoptOrphan(
+	ctx *pkgctx.VirtualMachineReplicaSetContext,
+	rs *vmopv1.VirtualMachineReplicaSet,
+	vm *vmopv1.VirtualMachine) error {
+
+	_, err := controllerutil.CreateOrPatch(ctx, r.Client, vm, func() error {
+		return controllerutil.SetControllerReference(rs, vm, r.Scheme())
```

Comment:
Do replica sets set ownership on _existing_ resources that have matching labels? The K8s replicaset controller _creates_ pods with the owner ref (https://github.com/kubernetes/kubernetes/blob/6d0aab2e38364d9d8e050b5f3a120a36ee588389/pkg/controller/replicaset/replica_set.go#L598), but I see no evidence it sets the owner ref on _adopted_ pods. Maybe I am missing something?

Ah, just found it. It's this bit in the replica set controller -- https://github.com/kubernetes/kubernetes/blob/6d0aab2e38364d9d8e050b5f3a120a36ee588389/pkg/controller/replicaset/replica_set.go#L756-L757. Which then leads to this in a PodControllerRefManager -- https://github.com/kubernetes/kubernetes/blob/6d0aab2e38364d9d8e050b5f3a120a36ee588389/pkg/controller/controller_ref_manager.go#L220-L234.

Reply:
Yep, same behavior with MachineSets adopting machines that match labels specified in the selector.